<a href="https://colab.research.google.com/github/KappaDistributive/Practical-Data-Science/blob/develop/Spam_Detection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Import Packages

In [0]:
import pandas as pd

# Obtain Data

Google Colab has a couple of nice features built-in that allows us to easily download our training data for spam detection from the author's website. Note that the following are Google Colab specific 'magic methdos' that won't work in any arbitrary Jupyter notebook.

In [15]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/00228/smsspamcollection.zip
!unzip smsspamcollection.zip
!mv SMSSpamCollection SMSSpamCollection.tsv
!mv readme readme.txt

--2020-02-20 11:59:13--  https://archive.ics.uci.edu/ml/machine-learning-databases/00228/smsspamcollection.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 203415 (199K) [application/x-httpd-php]
Saving to: ‘smsspamcollection.zip.1’

smsspamcollection.z 100%[===================>] 198.65K   996KB/s    in 0.2s    

2020-02-20 11:59:14 (996 KB/s) - ‘smsspamcollection.zip.1’ saved [203415/203415]

Archive:  smsspamcollection.zip
  inflating: SMSSpamCollection       
replace readme? [y]es, [n]o, [A]ll, [N]one, [r]ename: A
  inflating: readme                  


# Data Exploration

(Don't worry about the weird way in which we open and read the file here -- it turns out to be slightly corrupted.)

In [20]:
with open('readme.txt', 'rb') as f:
    print(f.read().decode('utf-8', 'ignore'))

SMS Spam Collection v.1
-------------------------

1. DESCRIPTION
--------------

The SMS Spam Collection v.1 (hereafter the corpus) is a set of SMS tagged messages that have been collected for SMS Spam research. It contains one set of SMS messages in English of 5,574 messages, tagged acording being ham (legitimate) or spam. 

1.1. Compilation
----------------

This corpus has been collected from free or free for research sources at the Web:

- A collection of between 425 SMS spam messages extracted manually from the Grumbletext Web site. This is a UK forum in which cell phone users make public claims about SMS spam messages, most of them without reporting the very spam message received. The identification of the text of spam messages in the claims is a very hard and time-consuming task, and it involved carefully scanning hundreds of web pages. The Grumbletext Web site is: http://www.grumbletext.co.uk/
- A list of 450 SMS ham messages collected from Caroline Tag's PhD Theses available 

According the the readme, our data contains 4827 non-spam SMS messagees and 747 spam SMS messages for a total of 5574 SMS messages.

The data is tab seperated, each line represents a data point with the first column containing the label and the second column containing the actual message.

Our first job is to verify this:

In [0]:
# Your code goes here:


In [46]:
#@title Sample Solution { display-mode: "form" }
raw_data = pd.read_csv('SMSSpamCollection.tsv', sep='\t', header=None, names=['label', 'message'])

# Your code goes here
print(f'Raw data shape: {raw_data.shape}\n')

print('Raw data head:')
print(raw_data.head())
print('\n')

print(f'Unique labels: {raw_data["label"].unique()}\n')

print(f'Label distribution: {raw_data["label"].value_counts()}\n')

Raw data shape: (5572, 2)

Raw data head:
  label                                            message
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...


Unique labels: ['ham' 'spam']

Label distribution: ham     4825
spam     747
Name: label, dtype: int64

